geomstats/geometry/sasaki_metric.py

```diff
@@ -0,0 +1,238 @@
+"""Class for the Sasaki metric.
+
+A class implementing the Sasaki metric: The natural metric on the tangent bundle TM
+of a Riemannian manifold M.
+
+Lead authors: E. Nava-Yazdani, F. Ambellan, M. Hanik and C. von Tycowicz.
+"""
+import os
+from joblib import Parallel, delayed
+
+import geomstats.backend as gs
+from geomstats.geometry.riemannian_metric import RiemannianMetric
+
+
+def _gradient_descent(x_ini, grad, exp, lrate=0.1, max_iter=100, tol=1e-6):
+    """Apply a gradient descent until max_iter or a given tolerance is reached."""
+    L = len(x_ini)
+    x = x_ini
+    for _ in range(max_iter):
+        grad_x = grad(x)
+        grad_norm = gs.linalg.norm(grad_x)
+        if grad_norm < tol:
+            break
+        grad_x = -lrate * grad_x
+        for j in range(L):
+            x[j] = exp(grad_x[j], x[j])
+    return list(x)
+
+
+class SasakiMetric(RiemannianMetric):
+    """Implements of the Sasaki metric on the tangent bundle TM of a Riem. manifold M.
+
+    The Sasaki metric is characterized by the following three properties:
+     * the canonical projection of TM becomes a Riemannian submersion,
+     * parallel vector fields along curves are orthogonal to their fibres, and
+     * its restriction to any tangent space is Euclidean.
+
+    Geodesic computations are realized via a discrete formulation of the geodesic
+    equation on TM that involve geodesics, parallel translation, and the curvature
+    tensor on the base manifold M (see [1] for details).
+
+    Parameters
+    ----------
+    metric : RiemannianMetric
+        Metric of the base manifold of the tangent bundle.
+    n_s : int
+        Number of steps used in time-discrete geodesic computations.
+        Optional. Default : 3.
+
+    References
+    ----------
+    .. [1] Muralidharan, P., & Fletcher, P. T. (2012, June).
+    Sasaki metrics for analysis of longitudinal data on manifolds.
+    In 2012 IEEE conference on computer vision and pattern recognition (pp. 1027-1034).
+    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4270017/
+    """
+
+    def __init__(self, metric: RiemannianMetric, n_s=3):
+        self.metric = metric  # Riemannian metric of underlying space
+        self.n_s = n_s  # Number of discretization steps
+        shape = (2, gs.prod(metric.shape))
```

Comment:
Well, this would introduce a difference if `metric.shape` consists of more than one element. I am not shure whether this should be changed as it my introduce problems.  Ok, I just checked back with @vontycowicz and he confirmed that `(2, *metric.shape)` did not work together with the Frechet-mean in Kendall shape space. For the moment we would like to keep it as is, unrolling `metric.shape`. This may be changed later on. 

Replacing `(2, gs.prod(metric.shape)` by `(2, gs.prod(gs.array(metric.shape))` already allows to successfully pass the unit tests employing the pytorch backend.

Reply:
Do you know why the unrolling hasn't worked with Frechet mean in Kendal shape space? Can you share a small reproducible example of the error?

The goal of unrolling would be that instead of reshaping to `(-1, 2) + self.metric.shape` later in the code, we could do `(-1,) + self.shape`. It would be more robust as we could control reshaping shape from one place only.

Reply:
@lpereira95, When taking `(2, *metric.shape)` instead of  `(2, gs.prod(metric.shape)` problems in the FrechetMean class at different locations, here with two examples:

1) In the constructor:
`        if point_type is None:
            self.point_type = metric.default_point_type
        error.check_parameter_accepted_values(
            self.point_type, "point_type", ["vector", "matrix"]
        )`
We would have something like 3-tensor as point_type and that is not handled.

2) In the _default_gradient_descent
`if point_type == "vector":
        points = gs.to_ndarray(points, to_ndim=2)
        einsum_str = "n,nj->j"
    else:
        points = gs.to_ndarray(points, to_ndim=3)
        einsum_str = "n,nij->ij"`
There is no handling for point_types different from vector and matrix. 

Apart form that, vectorizing the shape shape of the base manifold has the advantage that the iteration of taking the sasaki_metrix (i.e. looking at the tangent bundle of the tangent bundle) does not introduce new dimensions to the shape.